keps/sig-node/00014-resource-api.md

```diff
@@ -0,0 +1,425 @@
+
+## Background
+We are seeing increasing needs to better support non-native compute resources on Kubernetes that cover a wide range of resources such as GPUs, High-performance NICs, Infiniband and FPGAs. Such resources often require vendor specific setup, and have rich sets of different properties even across devices of the same type. This brings new requirements for Kubernetes to better support non-native compute resources to allow vendor specific setup, dynamic resource exporting, flexible resource configuration, and portable resource specification.
+
+The device plugin support added in Kubernetes 1.8 makes it easy for vendors to dynamically export their resources through a plugin API without changing Kubernetes core code. Taking a further step, this document proposes a new resource abstraction API, ResourceClass, that can be used to describe, manage and consume such vendor specific and metadata rich resources in simple and portable ways.
+
+## Use Stories
+### As a cluster operator:
+- Nodes in my cluster has GPU HW from different generations. I want to classify GPU nodes into one of the three categories, silver, gold and platinum depending upon the launch timeline of the GPU family eg: Kepler K20, K80, Pascal P40, P100, Volta V100. I want to charge each of the three categories differently. I want to offer my clients 3 GPU rates/classes to choose from.<br/>
+**Motivation:** As time progresses in a cluster lifecycle, new advanced, high performance, expensive variants of GPUs gets added to the cluster nodes. At the same time older variants also co-exist. There are workloads which strictly wants latest GPUs and also there are workloads which are fine with older GPUs. But since there is a wide range of types, it will be hard to manage and confusing at the same time to have granularity at each GPU type. Grouping into few broad categories will be convenient to manage.<br/>
+**Can this be solved without resource classes:** A unique taint can be used to represent a category like silver. Nodes can be tainted accordingly depending upon the type of GPUs availability. User pods can use tolerations to steer workloads to the appropriate nodes. But problem is how to restrict a user pod from not using the toleration that it should not be using?<br/>
+**How Resource classes can solve this:** I, operator/admin, creates three resource classes: GPU-Platinum, GPU-Gold, GPU-Silver. Now since resource classes are quota controlled, end-user will be able to request resource classes only if quota is allocated.
+
+- I want a mechanism where it is possible to offer a group of devices, which are co-located on a single node and shares a common property, as a single resource that can be requested in pod container spec. Example, N GPU units interconnected by NVLink or N cpu cores on same NUMA node.<br/>
+**Motivation:** Increased performance because of local access. Local access also helps better use of cache<br/>
```

Comment:
I thought we didn't want to bring topology into this at first?

More generally the motivation for using Resource Classes with topology is representing the topology and giving a way for the cluster admin to expose these shapes.

Resource Class won't increase performance.

Reply:
@RenaudWasTaken totally agree that ResourceClass itself won't increase performance. As mentioned earlier, what we are proposing here is an infrastructure building block through which we are hoping to fill some current gaps on non-primary resource scheduling and allow people to build more extensions to address their scheduling requirements in different environments. The original doc actually just stated such overall goals. However, during our f2f meeting in March, we got the feedback to provide more user stories. That is why Vikas and I wrote the "Resource Class User Stories" doc after the f2f meeting and presented that in the follow-up rwmg meeting to get consensus on the user story part. We included the user stories in this overall doc after getting positive feedback from rwmg meeting, but looks like so far, most of the discussions here are still on the user story part. That is really not our intention. What we are hoping is that people can review the overall design of this proposal and see whether it serves the purpose to enable their use cases as an infrastructure building block. The purpose of the user story part is just to provide some example use cases we heard from the community that we are hoping this design may help with. We never intend to say that this proposal alone can provide complete solution to all of these use cases. Especially for topology aware scheduling, I think another missing piece is to have a generic way of representing HW resource topology, which I think was also discussed during the f2f meeting.

Reply:
Totally agree with the "provide more user stories" part. 

> but looks like so far, most of the discussions here are still on the user story part. That is really not our intention. What we are hoping is that people can review the overall design of this proposal

I think the goal of user stories is to figure out what issue the system you are designing will solve. Agreeing on that will help getting the feature merged. My general impression here is that a lot of people have different expectations of the problem resource classes will help solve.

In any case my original comment was linked to the motivation here. "Increased performance because of local access" is not the motivation here.